### PR TITLE
[CI] Allow tests to be ran in parallel.

### PR DESCRIPTION
**Pull request by mandel-macaque**
This allows the CI to run ALL the tests that the project has in
parallel. This is divided in two main changes:

1. Xharness - We move away from using booleans to use a flag that states
   the tests to run.
2. yaml - We have move the code to use a template per label. This new
   jobs all run in parallel and the results are later collected by a
   funel job
3. pwsh - Added a new class that understands that we have several mark
   downs with the tests results. The classes parses them and them writes
   a single comment (and example can be found here: https://github.com/xamarin/xamarin-macios/pull/15201#issuecomment-1162366240

The changes gives the following advantages vs how we used to run tests:

1. The CI run for all tests moves from taking 13 hours to 3/4 hours
   (depending on the number of bots in the pool).
2. The download needed to verify the results on a case of failure is
   smaller. Rather than downloading several GBs we now just download
   that part of the html that we are interested in.
3. Better bot utlization. Bots are just used to a max of 2 hours, this
   means that we can use the bots better since they are fragmented.
4. Less VMs. VSDrops has added support for macOS and Linux, we take
   advanges of that here.

### CODE REVIEW

File: tools/devops/automation/scripts/TestResults.psm1 [hunk 1]
```diff
@@ -58,13 +74,265 @@ class TestResults {
...
+                                if (-not [string]::IsNullOrEmpty($line)) {
+                                    $stringBuilder.AppendLine("$line") # the extra space is needed for the multiline list item
+                                }
+                            }
+                        }
+                        $stringBuilder.AppendLine("</details>")
+                        $stringBuilder.AppendLine("")
+                    } else {
+                        $stringBuilder.AppendLine("<details>")
+                        $stringBuilder.AppendLine(" Test has no summaty file.")
```
Comment (rolfbjarne) [inline tools/devops/automation/scripts/TestResults.psm1:316]:
```suggestion
                        $stringBuilder.AppendLine(" Test has no summary file.")
```

File: tools/devops/automation/build-pull-request.yml [hunk 2]
```diff
@@ -56,33 +56,29 @@ parameters:
...
+    'linker',
+    'mac-binding-project',
+    'mmp',
+    'mononative',
+    'monotouch',
+    'msbuild',
+    'mtouch',
+    'xammac',
+    'xcframework',
+    'xtro'
```
[the remark this answers is not in the document]
Comment (mandel-macaque) [reply tools/devops/automation/build-pull-request.yml:81]: Variables do not support objects, but we can move the value to the template as the default value, and then not pass it around. The idea was to allow you to manually change the tests when you kick a build manually. Will give it a go.

**Comment by mandel-macaque** [comment]
mtouch tests failures will be fixed with https://github.com/xamarin/xamarin-macios/pull/15313

**Review by dalexsoto — APPROVED** [review]
❤️

**Pull request merged**